File: mcspring-build/mcspring-plugin-manager/src/main/java/in/kyle/mcspring/manager/commands/CommandPlugin.java
package in.kyle.mcspring.manager.commands;

import org.springframework.stereotype.Component;

import java.nio.file.Path;
import java.util.function.Function;
import java.util.stream.Collectors;

import in.kyle.mcspring.command.Command;
import in.kyle.mcspring.manager.controller.PluginController;
import in.kyle.mcspring.subcommands.PluginCommand;
import lombok.RequiredArgsConstructor;
import lombok.var;

@Component
@RequiredArgsConstructor
class CommandPlugin {
    
    private final PluginController pluginController;
    
    @Command(value = "plugin",
             aliases = "pl",
             description = "Load/unload/reload a specific plugin",
             usage = "/plugin <load|unload|list>")
    void plugin(PluginCommand command) {
        command.on("load", this::load);
        command.on("unload", this::unload);
        command.on("list", this::executeListPlugins);
        command.otherwise("Usage: /plugin <load|unload|list>");
    }
    
    private void load(PluginCommand command) {
        command.withMap(pluginController.getLoadablePlugins(),
                        s -> String.format("Plugin %s not found or is already loaded", s));
        command.then(this::executeLoad);
        command.otherwise("Usage: /plugin load <name>");
    }
    
    private void unload(PluginCommand command) {
        var plugins = pluginController.getPlugins()
                .stream()
                .collect(Collectors.toMap(org.bukkit.plugin.Plugin::getName, Function.identity()));
        command.withMap(plugins, s -> String.format("Plugin %s is not loaded", s));
        command.then(this::executeDisable);
        command.otherwise("Usage: /plugin unload <name>");
    }
    
    private String executeListPlugins() {
        return pluginController.getAllPlugins()
                .entrySet()
                .stream()
                .map(e -> String.format("%s%s", e.getValue() ? "&1" : "&4", e.getKey()))
                .collect(Collectors.joining(" "));
    }
    
    private String executeLoad(Path jar) {
        var pluginOptional = pluginController.load(jar);
        if (pluginOptional.isPresent()) {
            return String.format("Plugin %s enabled", pluginOptional.get().getName());
        } else {
            return String.format("Could not load %s see log for details", jar);
        }
    }
    
    private String executeDisable(org.bukkit.plugin.Plugin plugin) {
        boolean disabled = pluginController.unload(plugin);
        if (disabled) {
            return String.format("Plugin %s disabled", plugin.getName());
        } else {
            return String.format("Could not disable %s, see log for details", plugin.getName());
        }
    }
}

File: mcspring-build/mcspring-plugin-manager/src/main/java/in/kyle/mcspring/manager/commands/CommandPlugin.kt
package `in`.kyle.mcspring.manager.commands

import `in`.kyle.mcspring.command.Command
import `in`.kyle.mcspring.manager.controller.PluginController
import `in`.kyle.mcspring.subcommands.plugincommand.PluginCommand
import org.bukkit.plugin.Plugin
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean
import org.springframework.stereotype.Component
import java.nio.file.Path

@Component
@ConditionalOnBean(Plugin::class)
internal class CommandPlugin(
    private val pluginController: PluginController
) {

    @Command(value = "plugin", aliases = ["pl"], description = "Load/unload/reload a specific plugin", usage = "/plugin <load|unload|list>")
    fun plugin(command: PluginCommand) {
        command.on("load", ::load)
        command.on("unload", ::unload)
        command.on("list", ::executeListPlugins)
        command.otherwise("Usage: /plugin <load|unload|list>")
    }

    private fun load(command: PluginCommand) {
        command.withMap<Path>(pluginController.loadablePlugins) { "Plugin $it not found or is already loaded" }
        command.then(::executeLoad)
        command.otherwise("Usage: /plugin load <name>")
    }

    private fun unload(command: PluginCommand) {
        val plugins = pluginController.plugins.associateBy({ it.name }, { it })
        command.withMap(plugins) { "Plugin $it is not loaded" }
        command.then(::executeDisable)
        command.otherwise("Usage: /plugin unload <name>")
    }

    private fun executeListPlugins(): String {
        val colors = mapOf(true to "&1", false to "&2")
        return pluginController.allPlugins
                .map { "${colors[it.value]}${it.key}" }
                .joinToString(separator = " ")
    }

    private fun executeLoad(jar: Path): String {
        val plugin = pluginController.load(jar)

        return if (plugin != null) {
            "Plugin ${plugin.name} enabled"
        } else {
            "&4Could not load $jar see log for details"
        }
    }

    private fun executeDisable(plugin: Plugin): String {
        val disabled = pluginController.unload(plugin)
        return if (disabled) {
            "Plugin $plugin disabled"
        } else {
            "Could not disable $plugin, see log for details"
        }
    }
}
